# Reading and Writing Data

Real projects never start with a hand-typed DataFrame. Data lives in **files, databases and APIs**.
In this notebook we learn how to **load** data into a DataFrame and **save** it back.

**Contents**
1. The big picture: `read_*` and `to_*`
2. CSV (the most important one)
3. Excel
4. JSON (and nested JSON)
5. SQL databases (SQLite)
6. Parquet and other formats
7. Messy real-world files (the tricky parameters)
8. Big files: chunks, usecols, dtypes
9. File paths: the #1 beginner error
10. Pandas 3.x notes
11. Mini AI/ML example
12. Common mistakes
13. Interview questions
14. Practice questions with solutions
15. Cheat sheet

> **Connection with your earlier folders:** `11_file_handling`, `13_json`, `14_db_sqlite3`, `15_api_handling`. Earlier you read these files with plain Python (open, json, sqlite3, requests). Pandas does the same job in **one line** and gives you a DataFrame directly.

In [133]:
import numpy as np
import pandas as pd
from io import StringIO
from pathlib import Path
import sqlite3
import json

print("pandas:", pd.__version__)

# All sample files of this notebook will be created inside a "data" folder
DATA = Path("data")
DATA.mkdir(exist_ok=True)
print("data folder:", DATA.resolve())

pandas: 3.0.6
data folder: /Users/manishkumar/Documents/Explorer/CloudEng/MPython/27_pandas/theory/data


---
# 1. The big picture: `read_*` and `to_*`

Pandas has a **pair of functions for each format**:

| Format | Read (file -> DataFrame) | Write (DataFrame -> file) | Extra package needed |
|---|---|---|---|
| CSV | `pd.read_csv()` | `df.to_csv()` | none |
| Excel | `pd.read_excel()` | `df.to_excel()` | `openpyxl` |
| JSON | `pd.read_json()` | `df.to_json()` | none |
| SQL | `pd.read_sql()` | `df.to_sql()` | none for SQLite, `SQLAlchemy` + driver for PostgreSQL/MySQL |
| Parquet | `pd.read_parquet()` | `df.to_parquet()` | `pyarrow` |
| Pickle | `pd.read_pickle()` | `df.to_pickle()` | none |

Easy rule: **`read_` brings data IN, `to_` sends data OUT.**
```

---
# 2. CSV

**CSV** = Comma Separated Values. A plain text file where each line is a row and values are separated by commas. It is the most common data format in data science and ML (Kaggle datasets are mostly CSV).

## 2.1 First, create a sample file

(In real life you download the file. Here we create one so that the notebook runs on any computer.)

In [90]:
students_csv = """student_id,name,city,hours_studied,attendance,score
101,Ravi,Indore,5.5,80,85
102,Asha,Bhopal,7.0,90,92
103,Karan,Delhi,3.0,65,58
104,Neha,Indore,6.5,85,88
105,Amit,Pune,4.0,70,64
106,Sara,Bhopal,8.0,95,97
107,Dev,Delhi,2.5,60,45
108,Pooja,Indore,6.0,88,81
"""

(DATA / "students.csv").write_text(students_csv)
print("file created:", (DATA / "students.csv").exists())

file created: True


## 2.2 Reading a CSV

In [134]:
df = pd.read_csv("data/students.csv")
df

,student_id,name,city,hours_studied,attendance,score
0,101,Ravi,Indore,5.5,80,85
1,102,Asha,Bhopal,7.0,90,92
2,103,Karan,Delhi,3.0,65,58
3,104,Neha,Indore,6.5,85,88
4,105,Amit,Pune,4.0,70,64
5,106,Sara,Bhopal,8.0,95,97
6,107,Dev,Delhi,2.5,60,45
7,108,Pooja,Indore,6.0,88,81


In [92]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 8 entries, 0 to 7
Data columns (total 6 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   student_id     8 non-null      int64  
 1   name           8 non-null      str    
 2   city           8 non-null      str    
 3   hours_studied  8 non-null      float64
 4   attendance     8 non-null      int64  
 5   score          8 non-null      int64  
dtypes: float64(1), int64(3), str(2)
memory usage: 593.0 bytes


Pandas guessed the dtypes automatically: numbers became `int64`/`float64`, text became the string dtype.

## 2.3 Important `read_csv` parameters

| Parameter | Use | Example |
|---|---|---|
| `sep` | Separator if not comma | `sep=";"`, `sep="\t"` |
| `header` | Which row is the header (`None` = no header) | `header=None` |
| `names` | Give your own column names | `names=["a", "b"]` |
| `index_col` | Use a column as index | `index_col="student_id"` |
| `usecols` | Load only some columns (saves memory) | `usecols=["name", "score"]` |
| `nrows` | Load only first N rows (great for peeking at huge files) | `nrows=100` |
| `skiprows` | Skip lines at the top | `skiprows=2` |
| `dtype` | Force column types | `dtype={"zip": str}` |
| `na_values` | Extra strings to treat as NaN | `na_values=["?", "-"]` |
| `parse_dates` | Convert columns to datetime | `parse_dates=["date"]` |
| `thousands` | Handle numbers like `1,200` | `thousands=","` |
| `encoding` | File encoding | `encoding="utf-8"`, `"latin-1"` |
| `on_bad_lines` | What to do with broken rows | `on_bad_lines="skip"` |
| `chunksize` | Read in pieces (huge files) | `chunksize=10000` |

In [136]:
# Use student_id as the index
pd.read_csv("data/students.csv", index_col="student_id").head(3)

,name,city,hours_studied,attendance,score
student_id,,,,,
101,Ravi,Indore,5.5,80,85
102,Asha,Bhopal,7.0,90,92
103,Karan,Delhi,3.0,65,58


In [94]:
# Load only needed columns
pd.read_csv("data/students.csv", usecols=["name", "hours_studied", "score"]).head(3)

,name,hours_studied,score
0,Ravi,5.5,85
1,Asha,7.0,92
2,Karan,3.0,58


In [95]:
# Peek at only the first 3 rows
pd.read_csv("data/students.csv", nrows=3)

,student_id,name,city,hours_studied,attendance,score
0,101,Ravi,Indore,5.5,80,85
1,102,Asha,Bhopal,7.0,90,92
2,103,Karan,Delhi,3.0,65,58


In [96]:
# File without a header row -> give your own names
no_header = StringIO("""Ravi,21,85
Asha,22,92
""")
pd.read_csv(no_header, header=None, names=["name", "age", "score"])

,name,age,score
0,Ravi,21,85
1,Asha,22,92


In [97]:
# Different separator (semicolon, common in Europe) and tab-separated files
semi = StringIO("name;score\nRavi;85\nAsha;92")
print(pd.read_csv(semi, sep=";"))

   name  score
0  Ravi     85
1  Asha     92


## 2.4 Writing a CSV

Most important parameter: **`index=False`**. Without it, pandas saves the index as an extra useless column, and when you read the file again you get a strange `Unnamed: 0` column.

In [137]:
top = df[df["score"] >= 85]
top.to_csv("data/top_students.csv", index=False)

print(Path("data/top_students.csv").read_text())

student_id,name,city,hours_studied,attendance,score
101,Ravi,Indore,5.5,80,85
102,Asha,Bhopal,7.0,90,92
104,Neha,Indore,6.5,85,88
106,Sara,Bhopal,8.0,95,97



In [99]:
# What happens without index=False
top.to_csv("data/bad_example.csv")
print(pd.read_csv("data/bad_example.csv").head(2))   # see the Unnamed: 0 column

   Unnamed: 0  student_id  name    city  hours_studied  attendance  score
0           0         101  Ravi  Indore            5.5          80     85
1           1         102  Asha  Bhopal            7.0          90     92


In [100]:
# Other useful options
top.to_csv("data/top_small.csv", index=False, columns=["name", "score"])   # only some columns
top.to_csv("data/top_semi.csv", index=False, sep=";")                       # custom separator

---
# 3. Excel

Needs `openpyxl` (`pip install openpyxl`). One Excel file can have **many sheets**.

In [138]:
# Write one sheet
df.to_excel("data/students.xlsx", sheet_name="students", index=False)

# Read it back
pd.read_excel("data/students.xlsx", sheet_name="students").head(3)

,student_id,name,city,hours_studied,attendance,score
0,101,Ravi,Indore,5.5,80,85
1,102,Asha,Bhopal,7.0,90,92
2,103,Karan,Delhi,3.0,65,58


In [141]:
# Write MANY sheets into one file
by_city = df.groupby("city")["score"].mean().round(1).reset_index()

with pd.ExcelWriter("data/report.xlsx") as writer:
    df.to_excel(writer, sheet_name="students", index=False)
    by_city.to_excel(writer, sheet_name="city_average", index=False)

# Read all sheets at once -> dict {sheet_name: DataFrame}
sheets = pd.read_excel("data/report.xlsx", sheet_name=None)
print(list(sheets.keys()))
sheets["city_average"]

['students', 'city_average']


,city,score
0,Bhopal,94.5
1,Delhi,51.5
2,Indore,84.7
3,Pune,64.0


- `sheet_name="name"` or `sheet_name=0` -> one sheet
- `sheet_name=None` -> all sheets, returned as a dictionary
- Excel is slow and heavy for big data. For ML work prefer **CSV or Parquet**.

---
# 4. JSON

JSON is the format of **APIs** (you used it in `13_json` and `15_api_handling`). Two common shapes:

In [142]:
# Shape 1: list of records (most common from APIs)
records_json = """[
  {"name": "Ravi", "age": 21, "score": 85},
  {"name": "Asha", "age": 22, "score": 92}
]"""

pd.read_json(StringIO(records_json))

,name,age,score
0,Ravi,21,85
1,Asha,22,92


In [143]:
# Shape 2: dict of columns
cols_json = """{"name": ["Ravi", "Asha"], "score": [85, 92]}"""
pd.read_json(StringIO(cols_json))

,name,score
0,Ravi,85
1,Asha,92


## 4.1 Nested JSON (very common in real APIs)

Real API responses have dictionaries inside dictionaries. `read_json` cannot flatten them well. Use **`pd.json_normalize()`**.

In [105]:
api_data = [
    {"id": 1, "name": "Ravi", "address": {"city": "Indore", "pin": "452001"},
     "marks": {"maths": 85, "science": 90}},
    {"id": 2, "name": "Asha", "address": {"city": "Bhopal", "pin": "462001"},
     "marks": {"maths": 92, "science": 88}},
]

flat = pd.json_normalize(api_data)
flat

,id,name,address.city,address.pin,marks.maths,marks.science
0,1,Ravi,Indore,452001,85,90
1,2,Asha,Bhopal,462001,92,88


Nested keys became columns like `address.city` and `marks.maths`. You can rename them:

In [106]:
flat.columns = flat.columns.str.replace(".", "_", regex=False)
flat.columns.tolist()

['id', 'name', 'address_city', 'address_pin', 'marks_maths', 'marks_science']

In [107]:
# Nested list inside records -> use record_path
orders = [
    {"customer": "Ravi", "items": [{"product": "pen", "qty": 3}, {"product": "book", "qty": 1}]},
    {"customer": "Asha", "items": [{"product": "bag", "qty": 1}]},
]
pd.json_normalize(orders, record_path="items", meta=["customer"])

,product,qty,customer
0,pen,3,Ravi
1,book,1,Ravi
2,bag,1,Asha


## 4.2 Writing JSON

In [144]:
small = df[["name", "score"]].head(3)

print(small.to_json(orient="records", indent=2))

[
  {
    "name":"Ravi",
    "score":85
  },
  {
    "name":"Asha",
    "score":92
  },
  {
    "name":"Karan",
    "score":58
  }
]


In [145]:
small.to_json("data/small.json", orient="records", indent=2)
pd.read_json("data/small.json")

,name,score
0,Ravi,85
1,Asha,92
2,Karan,58


| `orient` | Output shape |
|---|---|
| `"records"` | `[{col: value}, ...]` (most used, API style) |
| `"columns"` (default) | `{col: {index: value}}` |
| `"index"` | `{index: {col: value}}` |
| `"split"` | `{"index": [...], "columns": [...], "data": [...]}` |

**API to DataFrame pattern** (from `15_api_handling`):
```python
import requests
response = requests.get(url)
df = pd.json_normalize(response.json())
```

---
# 5. SQL databases

You studied `sqlite3` in `14_db_sqlite3`. Pandas can read a **query result directly into a DataFrame** and write a DataFrame **into a table**.

In [148]:
conn = sqlite3.connect("data/school.db")

# DataFrame -> SQL table
df.to_sql("students", conn, if_exists="replace", index=False)
print("table written")

table written


In [149]:
# SQL query -> DataFrame
pd.read_sql("SELECT * FROM students WHERE score > 85", conn)

,student_id,name,city,hours_studied,attendance,score
0,102,Asha,Bhopal,7.0,90,92
1,104,Neha,Indore,6.5,85,88
2,106,Sara,Bhopal,8.0,95,97


In [112]:
# Aggregation done inside the database
query = """
SELECT city, COUNT(*) AS total, ROUND(AVG(score), 1) AS avg_score
FROM students
GROUP BY city
ORDER BY avg_score DESC
"""
pd.read_sql(query, conn)

,city,total,avg_score
0,Bhopal,2,94.5
1,Indore,3,84.7
2,Pune,1,64.0
3,Delhi,2,51.5


In [150]:
# SAFE way to pass values: use parameters (never build SQL with f-strings!)
pd.read_sql("SELECT name, score FROM students WHERE city = ? AND score > ?",
            conn, params=("Indore", 80))

,name,score
0,Ravi,85
1,Neha,88
2,Pooja,81


In [114]:
conn.close()

**`if_exists`** options of `to_sql`:
- `"fail"` (default): error if the table already exists
- `"replace"`: drop the table and create again
- `"append"`: add rows to the existing table

**For PostgreSQL / MySQL** (your `24_postgre_sql` folder) use SQLAlchemy:
```python
from sqlalchemy import create_engine
engine = create_engine("postgresql://user:password@localhost:5432/mydb")
df = pd.read_sql("SELECT * FROM students", engine)
df.to_sql("students_copy", engine, if_exists="append", index=False)
```
(Never hard-code real passwords in files you push to GitHub. Use environment variables or a `.env` file.)

**Good habit:** let the **database filter and aggregate** (WHERE, GROUP BY) when data is big, and bring only the needed result into pandas.

---
# 6. Parquet and other formats

**Parquet** is a **columnar binary format**. It is the standard in data engineering and big ML pipelines.

| | CSV | Parquet |
|---|---|---|
| Human readable | Yes | No |
| File size | Big | Small (compressed) |
| Read speed | Slow | Fast |
| Remembers dtypes | **No** (guesses again every time) | **Yes** |
| Read only some columns fast | No | Yes |

Needs `pip install pyarrow`.

In [151]:
try:
    df.to_parquet("data/students.parquet", index=False)
    back = pd.read_parquet("data/students.parquet")
    print(back.dtypes)
except ImportError:
    print("Install it first: pip install pyarrow")

student_id         int64
name                 str
city                 str
hours_studied    float64
attendance         int64
score              int64
dtype: object


In [152]:
# Size comparison on a bigger table
big = pd.DataFrame({
    "id": np.arange(200_000),
    "city": np.random.choice(["Indore", "Bhopal", "Delhi", "Pune"], 200_000),
    "score": np.random.randint(0, 100, 200_000),
})
big.to_csv("data/big.csv", index=False)

print("CSV size    :", round(Path("data/big.csv").stat().st_size / 1024), "KB")

try:
    big.to_parquet("data/big.parquet", index=False)
    print("Parquet size:", round(Path("data/big.parquet").stat().st_size / 1024), "KB")
except ImportError:
    print("pyarrow not installed")

CSV size    : 3045 KB
Parquet size: 1271 KB


**Pickle** (`to_pickle` / `read_pickle`) saves the DataFrame exactly as it is in Python. Fast, but only for Python and **never load pickle files from people you do not trust** (it can run code).

**Which format to choose?**
- Sharing with anyone / Kaggle / simple projects -> **CSV**
- Business people who use Excel -> **Excel**
- APIs, web apps -> **JSON**
- Data already in a database -> **SQL**
- Large data, ML pipelines, repeated reading -> **Parquet**

---
# 7. Messy real-world files

Real files are never clean. Look at this one: it has numbers with commas, strange missing markers, a date column, and an ID that starts with zeros.

In [153]:
messy = """order_id,pincode,customer,amount,order_date,rating
A1,00452,Ravi,"1,200",2026-01-05,4
A2,00462,Asha,"2,500",2026-01-06,?
A3,00110,Karan,-,2026-01-07,5
A4,00411,Neha,"3,100",2026-01-09,NA
"""

wrong = pd.read_csv(StringIO(messy))
print(wrong)
print()
print(wrong.dtypes)

  order_id  pincode customer amount  order_date rating
0       A1      452     Ravi  1,200  2026-01-05      4
1       A2      462     Asha  2,500  2026-01-06      ?
2       A3      110    Karan      -  2026-01-07      5
3       A4      411     Neha  3,100  2026-01-09    NaN

order_id        str
pincode       int64
customer        str
amount          str
order_date      str
rating          str
dtype: object


Problems with the default read:
1. `pincode` lost its leading zeros (`00452` became `452`)
2. `amount` is text because of `"1,200"` and `-`
3. `rating` is text because of `?`
4. `order_date` is plain text, not a real date

Fix everything with parameters:

In [118]:
good = pd.read_csv(
    StringIO(messy),
    dtype={"pincode": str},              # keep leading zeros
    thousands=",",                       # 1,200 -> 1200
    na_values=["?", "-", "NA"],          # these mean missing
    parse_dates=["order_date"],          # text -> datetime
)
print(good)
print()
print(good.dtypes)

  order_id pincode customer  amount order_date  rating
0       A1   00452     Ravi  1200.0 2026-01-05     4.0
1       A2   00462     Asha  2500.0 2026-01-06     NaN
2       A3   00110    Karan     NaN 2026-01-07     5.0
3       A4   00411     Neha  3100.0 2026-01-09     NaN

order_id                 str
pincode                  str
customer                 str
amount               float64
order_date    datetime64[us]
rating               float64
dtype: object


In [119]:
# Now we can do maths and date operations
print("Total amount:", good["amount"].sum())
print("Average rating (NaN skipped):", good["rating"].mean())
print(good["order_date"].dt.day_name().tolist())

Total amount: 6800.0
Average rating (NaN skipped): 4.5
['Monday', 'Tuesday', 'Wednesday', 'Friday']


## Broken rows

Some lines may have too many or too few columns (typical with scraped data):

In [120]:
broken = """name,score
Ravi,85
Asha,92,extra,values
Karan,70
"""

try:
    pd.read_csv(StringIO(broken))
except Exception as e:
    print("Error:", type(e).__name__)

print(pd.read_csv(StringIO(broken), on_bad_lines="skip"))

Error: ParserError
    name  score
0   Ravi     85
1  Karan     70


## Encoding errors

If you see `UnicodeDecodeError`, the file is not UTF-8 (common with Excel-exported CSV or Hindi/regional text). Try:
```python
pd.read_csv("file.csv", encoding="latin-1")
pd.read_csv("file.csv", encoding="utf-8-sig")   # CSV saved by Excel with BOM
```

---
# 8. Big files: chunks, usecols, dtypes

A 5 GB CSV will not fit in RAM. Three tricks:

1. **`usecols`**: load only the columns you need
2. **`dtype`**: use smaller types (`int32`, `category`)
3. **`chunksize`**: read piece by piece and combine results

In [154]:
# 1 + 2: usecols and category dtype
normal = pd.read_csv("data/big.csv")
light = pd.read_csv("data/big.csv", usecols=["city", "score"],
                    dtype={"city": "category", "score": "int8"})

print("normal :", round(normal.memory_usage(deep=True).sum() / 1024 ** 2, 2), "MB")
print("light  :", round(light.memory_usage(deep=True).sum() / 1024 ** 2, 2), "MB")

normal : 5.58 MB
light  : 0.38 MB


In [122]:
# 3: chunksize -> process the file piece by piece
total_rows = 0
score_sum = 0

for chunk in pd.read_csv("data/big.csv", chunksize=50_000):
    total_rows += len(chunk)
    score_sum += chunk["score"].sum()

print("rows   :", total_rows)
print("average:", round(score_sum / total_rows, 2))

rows   : 200000
average: 49.48


With `chunksize`, only 50,000 rows are in memory at a time, so even a huge file can be processed.

---
# 9. File paths: the #1 beginner error

`FileNotFoundError` is the most common error in this chapter. Reason: **a relative path is relative to where the notebook/script is run from**, not where the file is.

In [123]:
import os

print("Current working directory:", os.getcwd())
print("Files here:", os.listdir(".")[:8])

Current working directory: /Users/manishkumar/Documents/Explorer/CloudEng/MPython/27_pandas/theory
Files here: ['01_overview.ipynb', '03_reading_and_writing_data.ipynb', '0.0_pandas_yt.ipynb', '02_series_and_dataframe.ipynb', 'data', 'Test_data.csv']


Good habits:

```python
from pathlib import Path

# Always build paths with pathlib, works on Windows, Mac and Linux
data_path = Path("data") / "students.csv"
df = pd.read_csv(data_path)

# Check before reading
print(data_path.exists())
```

- Keep datasets in a `data/` folder inside the project.
- Add big data files to `.gitignore`, so they do not go to GitHub. (Small sample files are fine.)
- Never use full paths like `C:\Users\manish\...`; they break on any other computer.
- You can even read directly from a URL: `pd.read_csv("https://.../file.csv")`

---
# 10. Pandas 3.x notes

- Text columns are loaded with the dedicated **string dtype** (`str`), not `object`.
- Use `pd.read_json(StringIO(text))` for JSON text. Passing a raw JSON string directly was deprecated (as we did above).
- `to_csv`/`read_csv` behave the same as before, but always check `df.dtypes` after loading, especially for IDs, pincodes and dates.

---
# 11. Mini AI/ML example: load -> clean -> save -> reload

The standard ML data flow: read the raw file, clean it, save the **clean** version in a separate file, and the model code reads only the clean file.

In [155]:
raw = StringIO("""student_id,hours_studied,attendance,score,city
1,5.5,80,85,Indore
2,7,,92,Bhopal
3,3,65,58,Delhi
3,3,65,58,Delhi
4,6.5,85,?,Indore
5,4,70,64,Pune
""")

# 1. Load (tell pandas "?" means missing)
raw_df = pd.read_csv(raw, na_values=["?"])
print("raw shape:", raw_df.shape)

# 2. Clean
clean = (
    raw_df
    .drop_duplicates()
    .assign(attendance=lambda d: d["attendance"].fillna(d["attendance"].median()))
    .dropna(subset=["score"])                  # no score = cannot learn from this row
)
print("clean shape:", clean.shape)
clean

raw shape: (6, 5)
clean shape: (4, 5)


,student_id,hours_studied,attendance,score,city
0,1,5.5,80.0,85.0,Indore
1,2,7.0,75.0,92.0,Bhopal
2,3,3.0,65.0,58.0,Delhi
5,5,4.0,70.0,64.0,Pune


In [125]:
# 3. Save the clean version
clean.to_csv("data/students_clean.csv", index=False)

# 4. Model code reads the clean file later
X = pd.read_csv("data/students_clean.csv")[["hours_studied", "attendance"]].to_numpy()
y = pd.read_csv("data/students_clean.csv")["score"].to_numpy()
print(X.shape, y.shape)

(4, 2) (4,)


---
# 12. Common mistakes

| Mistake | Fix |
|---|---|
| `to_csv()` without `index=False` -> `Unnamed: 0` column later | `to_csv(path, index=False)` |
| `FileNotFoundError` | Check `os.getcwd()`, use `pathlib`, check `.exists()` |
| ID / pincode / phone loses leading zeros | `dtype={"pincode": str}` |
| Numbers stay as text because of `,` or `-` | `thousands=","`, `na_values=["-"]` |
| Date column stays as text | `parse_dates=["date"]` |
| `read_excel` fails with ImportError | `pip install openpyxl` |
| `UnicodeDecodeError` | Try `encoding="latin-1"` or `"utf-8-sig"` |
| Reading a huge file fully | `usecols`, `dtype`, `chunksize`, or Parquet |
| Building SQL with f-strings (SQL injection risk) | Use `params=` |
| Nested JSON gives dict objects inside cells | `pd.json_normalize()` |
| Pushing big data files or passwords to GitHub | `.gitignore`, `.env` |

---
# 13. Interview questions

**Q1. How do you read a large CSV that does not fit in memory?**
Use `chunksize` to process in pieces, `usecols` to load only needed columns, smaller `dtype`s (`category`, `int32`), or switch to Parquet / a database / Polars / Spark.

**Q2. Why do we use `index=False` in `to_csv`?**
Otherwise the index is written as an extra column, and it comes back as `Unnamed: 0` when the file is read again.

**Q3. What is `json_normalize` and when is it used?**
It flattens nested JSON (dicts/lists inside records) into a flat table. Used for API responses.

**Q4. CSV vs Parquet?**
CSV is text, human readable, big and slow, and does not store dtypes. Parquet is binary, columnar, compressed, fast, and keeps dtypes. Parquet is better for large data and pipelines.

**Q5. How do you handle values like "?" or "N/A" in a CSV?**
`na_values=["?", "N/A"]` while reading, so they become NaN.

**Q6. How do you keep leading zeros in a column like pincode?**
Read it as a string: `dtype={"pincode": str}`.

**Q7. How do you load only specific columns?**
`usecols=[...]`. It saves memory and time.

**Q8. How do you read from a SQL database?**
`pd.read_sql(query, connection)`. Use `params` for values to avoid SQL injection. For PostgreSQL/MySQL use a SQLAlchemy engine.

**Q9. What does `if_exists` do in `to_sql`?**
Controls behaviour when the table exists: `fail`, `replace` or `append`.

**Q10. Why can pickle be dangerous?**
Unpickling untrusted data can execute arbitrary code. Only load pickle files you created yourself.

---
# 14. Practice questions

Use this raw data for all questions (run the cell first).

In [126]:
practice_csv = """emp_id;name;dept;salary;join_date
E01;Amit;ML;"1,20,000";2023-04-10
E02;Bina;Web;95000;2022-11-01
E03;Chirag;ML;?;2024-01-15
E04;Divya;Data;"1,10,000";2021-07-20
E05;Esha;Web;88000;2023-09-05
"""
Path("data/employees.csv").write_text(practice_csv)

192

**Q1.** Read `data/employees.csv` correctly (note the `;` separator). Print the dtypes.
**Q2.** Fix it: `?` should be missing, `salary` should be numeric (the commas are Indian-style `1,20,000`), `join_date` should be a datetime, and `emp_id` should be the index.
**Q3.** Load only `name` and `dept` columns.
**Q4.** Save the fixed DataFrame (from Q2) to `data/employees_clean.csv` without an extra index column, then read it back and confirm the shape is the same.
**Q5.** Write the fixed DataFrame into a SQLite table `employees`, then use `read_sql` to get average salary per department.
**Q6.** Flatten this nested JSON into a DataFrame:
```python
data = [{"id": 1, "user": {"name": "Ravi", "city": "Indore"}},
        {"id": 2, "user": {"name": "Asha", "city": "Bhopal"}}]
```

### Solutions

In [127]:
# Q1
q1 = pd.read_csv("data/employees.csv", sep=";")
print(q1.dtypes)
q1

emp_id       str
name         str
dept         str
salary       str
join_date    str
dtype: object


,emp_id,name,dept,salary,join_date
0,E01,Amit,ML,"1,20,000",2023-04-10
1,E02,Bina,Web,95000,2022-11-01
2,E03,Chirag,ML,?,2024-01-15
3,E04,Divya,Data,"1,10,000",2021-07-20
4,E05,Esha,Web,88000,2023-09-05


In [128]:
# Q2
emp = pd.read_csv(
    "data/employees.csv",
    sep=";",
    na_values=["?"],
    thousands=",",
    parse_dates=["join_date"],
    index_col="emp_id",
)
print(emp.dtypes)
emp

name                    str
dept                    str
salary              float64
join_date    datetime64[us]
dtype: object


,name,dept,salary,join_date
emp_id,,,,
E01,Amit,ML,120000.0,2023-04-10
E02,Bina,Web,95000.0,2022-11-01
E03,Chirag,ML,NaN,2024-01-15
E04,Divya,Data,110000.0,2021-07-20
E05,Esha,Web,88000.0,2023-09-05


In [129]:
# Q3
pd.read_csv("data/employees.csv", sep=";", usecols=["name", "dept"])

,name,dept
0,Amit,ML
1,Bina,Web
2,Chirag,ML
3,Divya,Data
4,Esha,Web


In [130]:
# Q4
emp.to_csv("data/employees_clean.csv", index=False)
again = pd.read_csv("data/employees_clean.csv")
print(emp.shape, again.shape)
# Note: we used index=False, so emp_id (the index) was not saved.
# To keep it: emp.to_csv(path) and read with index_col="emp_id"

(5, 4) (5, 4)


In [131]:
# Q5
conn = sqlite3.connect("data/company.db")
emp.reset_index().to_sql("employees", conn, if_exists="replace", index=False)

print(pd.read_sql("SELECT dept, ROUND(AVG(salary), 0) AS avg_salary FROM employees GROUP BY dept", conn))
conn.close()

   dept  avg_salary
0  Data    110000.0
1    ML    120000.0
2   Web     91500.0


In [132]:
# Q6
data = [{"id": 1, "user": {"name": "Ravi", "city": "Indore"}},
        {"id": 2, "user": {"name": "Asha", "city": "Bhopal"}}]
pd.json_normalize(data)

,id,user.name,user.city
0,1,Ravi,Indore
1,2,Asha,Bhopal


---
# 15. Cheat sheet

```python
# CSV
df = pd.read_csv("data/file.csv", sep=",", index_col=None, usecols=None, nrows=None,
                 dtype=None, na_values=None, parse_dates=None, thousands=None,
                 encoding="utf-8", on_bad_lines="error", chunksize=None)
df.to_csv("data/out.csv", index=False)

# Excel
df = pd.read_excel("file.xlsx", sheet_name=0)        # None = all sheets (dict)
df.to_excel("out.xlsx", sheet_name="data", index=False)
with pd.ExcelWriter("report.xlsx") as w:             # many sheets
    df1.to_excel(w, sheet_name="a", index=False)

# JSON
df = pd.read_json(StringIO(text))                    # or a file path
df = pd.json_normalize(nested_list)                  # nested JSON
df.to_json("out.json", orient="records", indent=2)

# SQL
df = pd.read_sql("SELECT ...", conn, params=(value,))
df.to_sql("table", conn, if_exists="append", index=False)

# Parquet / Pickle
df.to_parquet("out.parquet");  pd.read_parquet("out.parquet")

# Paths
from pathlib import Path
Path("data") / "file.csv";  Path("data").mkdir(exist_ok=True)
```

**Next chapter:** `04_inspecting_data`: `head`, `info`, `describe`, `value_counts`, `nunique`, `unique`, and how to understand any new dataset in 5 minutes.